app/app_main.c

```diff
@@ -2583,52 +2639,55 @@ static ACVP_RESULT app_aes_handler_aead(ACVP_TEST_CASE *test_case)
-            EVP_CIPHER_CTX_ctrl(&cipher_ctx, EVP_CTRL_CCM_SET_TAG, tc->tag_len, 0);
-            EVP_CipherInit(&cipher_ctx, NULL, tc->key, tc->iv, 1);
-            EVP_Cipher(&cipher_ctx, NULL, NULL, tc->pt_len);
-            EVP_Cipher(&cipher_ctx, NULL, tc->aad, tc->aad_len);
-            EVP_Cipher(&cipher_ctx, tc->ct, tc->pt, tc->pt_len);
-            EVP_CIPHER_CTX_ctrl(&cipher_ctx, EVP_CTRL_CCM_GET_TAG, tc->tag_len, tc->ct + tc->ct_len);
+            EVP_CipherInit(cipher_ctx, cipher, NULL, NULL, 1);
+            EVP_CIPHER_CTX_ctrl(cipher_ctx, EVP_CTRL_CCM_SET_IVLEN, tc->iv_len, 0);
+            EVP_CIPHER_CTX_ctrl(cipher_ctx, EVP_CTRL_CCM_SET_TAG, tc->tag_len, 0);
+            EVP_CipherInit(cipher_ctx, NULL, tc->key, tc->iv, 1);
+            EVP_Cipher(cipher_ctx, NULL, NULL, tc->pt_len);
+            EVP_Cipher(cipher_ctx, NULL, tc->aad, tc->aad_len);
+            EVP_Cipher(cipher_ctx, tc->ct, tc->pt, tc->pt_len);
+            EVP_CIPHER_CTX_ctrl(cipher_ctx, EVP_CTRL_CCM_GET_TAG, tc->tag_len, tc->ct + tc->ct_len);
             tc->ct_len += tc->tag_len;
         } else if (tc->direction == ACVP_DIR_DECRYPT) {
-          EVP_CipherInit(&cipher_ctx, cipher, NULL, NULL, 0);
-          EVP_CIPHER_CTX_ctrl(&cipher_ctx, EVP_CTRL_CCM_SET_IVLEN, tc->iv_len, 0);
-          EVP_CIPHER_CTX_ctrl(&cipher_ctx, EVP_CTRL_CCM_SET_TAG, tc->tag_len, tc->ct + tc->pt_len);
-          EVP_CipherInit(&cipher_ctx, NULL, tc->key, tc->iv, 0);
-          EVP_Cipher(&cipher_ctx, NULL, NULL, tc->pt_len);
-          EVP_Cipher(&cipher_ctx, NULL, tc->aad, tc->aad_len);
-          /*
-           * Decrypt and check the tag
-           */
-          rv = EVP_Cipher(&cipher_ctx, tc->pt, tc->ct, tc->pt_len);
-          if (rv < 0) {
-              EVP_CIPHER_CTX_cleanup(&cipher_ctx);
-              return ACVP_CRYPTO_TAG_FAIL;
-          }
+            EVP_CipherInit(cipher_ctx, cipher, NULL, NULL, 0);
+            EVP_CIPHER_CTX_ctrl(cipher_ctx, EVP_CTRL_CCM_SET_IVLEN, tc->iv_len, 0);
+            EVP_CIPHER_CTX_ctrl(cipher_ctx, EVP_CTRL_CCM_SET_TAG, tc->tag_len, tc->ct + tc->pt_len);
+            EVP_CipherInit(cipher_ctx, NULL, tc->key, tc->iv, 0);
+            EVP_Cipher(cipher_ctx, NULL, NULL, tc->pt_len);
+            EVP_Cipher(cipher_ctx, NULL, tc->aad, tc->aad_len);
+            /*
+             * Decrypt and check the tag
+             */
+            ret = EVP_Cipher(cipher_ctx, tc->pt, tc->ct, tc->pt_len);
+            if (ret < 0) {
+                rc = ACVP_CRYPTO_TAG_FAIL;
+                goto end;
+            }
         }
         break;
     default:
         printf("Error: Unsupported AES AEAD mode requested by ACVP server\n");
-            EVP_CIPHER_CTX_cleanup(&cipher_ctx);
-            return ACVP_NO_CAP;
+        rc = ACVP_NO_CAP;
+        goto end;
     }
 
-    EVP_CIPHER_CTX_cleanup(&cipher_ctx);
+end:
+    /* Cleanup */
+    if (cipher_ctx) EVP_CIPHER_CTX_free(cipher_ctx);
```

Comment:
Yes! Thank you... a similar thing was causing a big memory leak in the cmac code previously :)